packages/react/src/ReactElement.js

```diff
@@ -321,23 +324,7 @@ export function cloneElement(element, config, children) {
     }
 
     // Remaining properties override existing props
-    let defaultProps;
-    if (element.type && element.type.defaultProps) {
-      defaultProps = element.type.defaultProps;
-    }
-    for (propName in config) {
```

Comment:
I am a bit confuse and worry that these lines will not produce the same result as `setProps` function do?

Reply:
This is where I initially thought they were doing something different as well. This code was actually added later to fix a bug with default props (undefined values were not falling back to default props) that only existed in `cloneElement` via https://github.com/facebook/react/pull/5997/files.

The real difference structurally between how it was and how it is in this PR is when default props are set (if needed). Currently they are set while setting props, but before setting children on props. The order ultimately doesn't matter because default props will only be set when a value is undefined and clone element. We can also see from the tests passing that this functionality still behaves as expected.

Though actually, from playing with this more I this difference is an inconsistency when `children` is explicitly set to `undefined` and the component has a default prop for `children`. See the issue I just opened: https://github.com/facebook/react/issues/14136.

I'll add a test case depending on the outcome of that issue.